lib/kcerts/ssh.go

```diff
@@ -0,0 +1,199 @@
+package kcerts
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/cache"
+	"github.com/enfabrica/enkit/lib/logger/klog"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+var (
+	sockR = regexp.MustCompile("(?m)SSH_AUTH_SOCK=([^;\\n]*)")
+	pidR  = regexp.MustCompile("(?m)SSH_AGENT_PID=([0-9]*)")
+)
+
+// FindSSHDir will find the users ssh directory based on $HOME. If $HOME/.ssh does not exist
+// it will attempt to create it.
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	defer knownHostsFile.Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type SSHAgent struct {
+	PID    int    `json:"pid"`
+	Socket string `json:"sock"`
+	// failed is marked in WriteToCache
+	failed bool
+}
+
+func (a SSHAgent) Valid() bool {
+	conn, err := net.Dial("unix", a.Socket)
+	defer conn.Close()
+	return err == nil
+}
+
+// CloseIfNotCached will kill the agent if it failed to write to cache. Otherwise is a NoOp
+func (a *SSHAgent) CloseIfNotCached() error {
+	if a.failed {
+		cmd := exec.Command("kill", "-HUP", strconv.Itoa(a.PID))
+		cmd.Stdout = os.Stdout
```

Comment:
If you follow my advice of creating a Close attribute above, CloseIfNotFailed can be removed entirely, get rid of it.
You'll be left with: `Close` attribute (that depending on how the agent was created it'll close it in different ways) and `Kill`, to actually kill the agent.